testament/backend.nim

```diff
@@ -37,34 +37,109 @@ proc getCommit(): CommitId =
   result = CommitId(hash)
 
 var
-  results: File
+  results: JsonNode
   currentCategory: string
-  entries: int
 
 proc writeTestResult*(name, category, target, action, result, expected, given: string) =
   createDir("testresults")
   if currentCategory != category:
     if currentCategory.len > 0:
-      results.writeLine("]")
-      close(results)
+      var resFile = open("testresults" / category.addFileExt"json", fmWrite)
+      resFile.write pretty(results)
+      close resFile
     currentCategory = category
-    results = open("testresults" / category.addFileExt"json", fmWrite)
-    results.writeLine("[")
-    entries = 0
+    results = newJArray()
 
-  let jentry = %*{"name": name, "category": category, "target": target,
+  results.add %*{"name": name, "category": category, "target": target,
     "action": action, "result": result, "expected": expected, "given": given,
     "machine": thisMachine.string, "commit": thisCommit.string, "branch": thisBranch}
-  if entries > 0:
-    results.writeLine(",")
-  results.write($jentry)
-  inc entries
 
 proc open*() =
   thisMachine = getMachine()
   thisCommit = getCommit()
 
 proc close*() =
   if currentCategory.len > 0:
-    results.writeLine("]")
-    close(results)
+    var resFile = open("testresults" / currentCategory.addFileExt"json", fmWrite)
+    resFile.write pretty(results)
+    close resFile
+
+proc cacheResults*() =
+  ## Traverses the testresults directory and extracts any failed json entries
+  ## and inserts them into a new file within the cacheresults directory.
+  createDir("testresults/cacheresults")
+  # We will ignore any entries that have this as their result
+  const passResults = ["reJoined", "reSuccess", "reDisabled", ""] # "" is defaulted to if result field not found
+  let searchPattern = "testresults" / "*.json"
+  # Prepare json array which will be written to our new cache file
+  var fresults = newJArray()
+  # Walk files in testresults that have a json ext
+  for f in walkFiles(searchPattern):
+    var jdata: JsonNode
+    try:
+      # Parse the json file into the json node
+        jdata = f.parseFile()
+    except IOError, JsonParsingError:
+      echo getCurrentExceptionMsg()
```

Comment:
You need to do better than just printing the exception message. Don't print an error and pretend like nothing happened.

Reply:
Yep thats my bad; had changed the other occurrence to have a better message printed but didn't copy it over